src/orion/algo/space/configspace.py

```diff
@@ -0,0 +1,261 @@
+        raise NotImplementedError()
+
+    def real(self, dim: Real) -> Optional[FloatHyperparameter]:
+        """Convert a real dimension into a configspace equivalent"""
+        if dim.prior_name in ("reciprocal", "uniform"):
+            a, b = dim._args
+
+            return UniformFloatHyperparameter(
+                name=dim.name,
+                lower=a,
+                upper=b,
+                default_value=dim.default_value,
+                q=_qantization(dim),
+                log=dim.prior_name == "reciprocal",
+            )
+
+        if dim.prior_name in ("normal", "norm"):
+            a, b = dim._args
+
+            # TODO: #845 remove python 3.6 special case
+            kwargs = dict(
+                name=dim.name,
+                mu=a,
+                sigma=b,
+                default_value=dim.default_value,
+                q=_qantization(dim),
+                log=False,
+                lower=dim.low if hasattr(dim, "low") else None,
+                upper=dim.high if hasattr(dim, "high") else None,
+            )
+
+            return NormalFloatHyperparameter(**kwargs)
+
+        return
+
+    def integer(self, dim: Integer) -> Optional[IntegerHyperparameter]:
+        """Convert a integer dimension into a configspace equivalent"""
+        if dim.prior_name in ("int_uniform", "int_reciprocal"):
+            a, b = dim._args
+
+            return UniformIntegerHyperparameter(
+                name=dim.name,
+                lower=a,
+                upper=b,
+                default_value=dim.default_value,
+                q=_qantization(dim),
+                log=dim.prior_name == "int_reciprocal",
+            )
+
+        if dim.prior_name in ("norm", "normal"):
+            a, b = dim._args
+
+            # TODO: #845 remove python 3.6 special case
+            kwargs = dict(
+                name=dim.name,
+                mu=a,
+                sigma=b,
+                default_value=dim.default_value,
+                q=_qantization(dim),
+                log=False,
+                lower=dim.low if hasattr(dim, "low") else None,
+                upper=dim.high if hasattr(dim, "high") else None,
+            )
+
+            return NormalIntegerHyperparameter(**kwargs)
+
+        return None
+
+    def categorical(self, dim: Categorical) -> Optional[CategoricalHyperparameter]:
+        """Convert a categorical dimension into a configspace equivalent"""
+        return CategoricalHyperparameter(
+            name=dim.name,
+            choices=dim.categories,
+            weights=dim._probs,
+        )
+
+    def fidelity(self, dim: Fidelity) -> None:
+        """Ignores fidelity dimension as configspace does not have an equivalent"""
+        return None
+
+    def space(self, space: Space) -> ConfigurationSpace:
+        """Convert orion space to configspace"""
+        cspace = ConfigurationSpace()
+        dims = []
+
+        for _, dim in space.items():
```

Comment:
I think this part could be a generic convert_dimensions method that we do not need to reimplement for each visitors. On a similar topic, I know this deviates from the Visitor pattern but I think a single method `convert_dimension` like 
```
class Visitor:
    ...
    def convert_dimension(self, dimension: Dimension)
        return getattr(self, _to_snake_case(dimension.__name__))(dimension)
```
Also, maybe a class name along the lines of `DimensionConverter` or `SpaceConverter` would be more explicit. To be fair, this would start to be very close to singledispatch. I agree however that the class structure is helpful for contributors to understand what they need to cover.

Reply:
But a visitor is not about converting the space, we are using it to do so, but it is a completely generic abstraction that has nothing to do with converting the dimension, that is why the name of the methods are non specific because it is just about visiting the tree nodes.

You could use the visitor to implement any of the method defined inside the Dimension class and subclasses

Reply:
I'm worried that we are using an abstraction that will not be used for something else than conversions and that makes the implementation of conversions less intuitive for external contributors. If there is a relatively clear potential for other usages of this abstraction then I think it's fine, but right now I don't see any. What would you have in mind for other uses?

Reply:
updated

Reply:
If not using the visitor pattern then I think there is no need to adding the `convert_dimension` to Dimension, this could be part of the base SpaceConverter.

Reply:
it already does not have it

Reply:
Oh right, I missred the diff. Sorry about that!